src/transformers/pipelines/__init__.py

```diff
@@ -434,6 +433,65 @@ def pipeline(
             model, config=config, revision=revision, _from_pipeline=task, **model_kwargs
         )
 
+    model_config = model.config
+
+    load_tokenizer = type(model_config) in TOKENIZER_MAPPING
+    load_feature_extractor = type(model_config) in FEATURE_EXTRACTOR_MAPPING
+
+    if load_tokenizer:
+        # Try to infer tokenizer from model or config name (if provided as str)
+        if tokenizer is None:
+            if isinstance(model_name, str):
+                tokenizer = model_name
+            elif isinstance(config, str):
+                tokenizer = config
+            else:
+                # Impossible to guess what is the right tokenizer here
+                raise Exception(
+                    "Impossible to guess which tokenizer to use. "
+                    "Please provide a PreTrainedTokenizer class or a path/identifier to a pretrained tokenizer."
+                )
+
+        # Instantiate tokenizer if needed
+        if isinstance(tokenizer, (str, tuple)):
+            if isinstance(tokenizer, tuple):
+                # For tuple we have (tokenizer name, {kwargs})
+                use_fast = tokenizer[1].pop("use_fast", use_fast)
+                tokenizer_identifier = tokenizer[0]
+                tokenizer_kwargs = tokenizer[1]
+            else:
+                tokenizer_identifier = tokenizer
+                tokenizer_kwargs = model_kwargs
+
+            tokenizer = AutoTokenizer.from_pretrained(
+                tokenizer_identifier, revision=revision, use_fast=use_fast, _from_pipeline=task, **tokenizer_kwargs
+            )
+    else:
+        tokenizer = None
```

Comment:
That whole logic seems extremely convoluted and error prone.

- You override the `tokenizer` to None if `load_tokenizer == False`. That's a very bad behavior for this function in my opinion. If I pass you an already instantiated class, you should just pass it as it, you should never override it (Imagine someone implemented their own tokenizer)
- `tokenizer_kwargs = model_kwargs` seems super risky. They are not the same, this is bound to cause some errors at some point.
- We could probably make the intent easier to understand too by using different names. for instance like `model_name` is necessarily a string we could have `tokenizer_name` (or `tokenizer_identifier`) 
- `tokenizer` argument should never be a tuple (it's not stated in the function signature at least, and it seems error prone) (This was already there, so we should keep it, but still a wart no ?)

Reply:
You're right about overriding the tokenizer to `None` - the rest were already there in some shape or form but happy to cleanup in this PR. Will propose changes.

Reply:
Yes I know some stuff was already there. The biggest point as you noted is silent override of the argument.

This function is a challenge to modify ;)

Reply:
In the end I modified the `tokenizer` set to `None`, and left the rest as it initially was. As said above, happy to tackle this in a cleanup PR.

I added a test to ensure that we could pass a custom tokenizer per the use-case you've mentioned.